testing/endtoend/e2e_crash_test.go

```diff
@@ -0,0 +1,303 @@
+	require.NoError(t, err)
+	dataHash := common.Hash{1}
+	enqueueSequencerMessageAsExecutor(
+		t, accounts[0], rollupAddr.UpgradeExecutor, bk.Client(), bridgeAddr, seqMessage{
+			dataHash:                 dataHash,
+			afterDelayedMessagesRead: big.NewInt(1),
+			prevMessageCount:         big.NewInt(1),
+			newMessageCount:          big.NewInt(2),
+		},
+	)
+
+	baseStateManagerOpts := []statemanager.Opt{
+		statemanager.WithNumBatchesRead(inboxCfg.numBatchesPosted),
+		statemanager.WithLayerZeroHeights(&protocolCfg.layerZeroHeights, protocolCfg.numBigStepLevels),
+	}
+	honestStateManager, err := statemanager.NewForSimpleMachine(t, baseStateManagerOpts...)
+	require.NoError(t, err)
+
+	shp := &simpleHeaderProvider{b: bk, chs: make([]chan<- *gethtypes.Header, 0)}
+	shp.Start(neutralCtx)
+
+	baseStackOpts := []cm.StackOpt{
+		cm.StackWithMode(types.MakeMode),
+		cm.StackWithPollingInterval(timeCfg.assertionScanningInterval),
+		cm.StackWithPostingInterval(timeCfg.assertionPostingInterval),
+		cm.StackWithAverageBlockCreationTime(timeCfg.blockTime),
+		cm.StackWithConfirmationInterval(timeCfg.assertionConfirmationAttemptInterval),
+		cm.StackWithHeaderProvider(shp),
+	}
+
+	name := "honest"
+	txOpts := accounts[1]
+	//nolint:gocritic
+	honestOpts := append(
+		baseStackOpts,
+		cm.StackWithName(name),
+	)
+	honestChain := setupAssertionChain(t, honestCtx, bk.Client(), rollupAddr.Rollup, txOpts)
+	honestManager, err := cm.NewChallengeStack(honestChain, honestStateManager, honestOpts...)
+	require.NoError(t, err)
+
+	totalOpcodes := totalWasmOpcodes(&protocolCfg.layerZeroHeights, protocolCfg.numBigStepLevels)
+	t.Logf("Total wasm opcodes in test: %d", totalOpcodes)
+
+	assertionDivergenceHeight := uint64(1)
+	assertionBlockHeightDifference := int64(1)
+
+	machineDivergenceStep := uint64(1)
+	//nolint:gocritic
+	evilStateManagerOpts := append(
+		baseStateManagerOpts,
+		statemanager.WithMachineDivergenceStep(machineDivergenceStep),
+		statemanager.WithBlockDivergenceHeight(assertionDivergenceHeight),
+		statemanager.WithDivergentBlockHeightOffset(assertionBlockHeightDifference),
+	)
+	evilStateManager, err := statemanager.NewForSimpleMachine(t, evilStateManagerOpts...)
+	require.NoError(t, err)
+
+	// Honest validator has index 1 in the accounts slice, as 0 is admin, so
+	// evil ones should start at 2.
+	evilTxOpts := accounts[2]
+	//nolint:gocritic
+	evilOpts := append(
+		baseStackOpts,
+		cm.StackWithName("evil"),
+	)
+	evilChain := setupAssertionChain(t, evilCtx, bk.Client(), rollupAddr.Rollup, evilTxOpts)
+	evilManager, err := cm.NewChallengeStack(evilChain, evilStateManager, evilOpts...)
+	require.NoError(t, err)
+
+	chalManagerAddr := honestChain.SpecChallengeManager().Address()
+	cmBindings, err := challengeV2gen.NewEdgeChallengeManager(chalManagerAddr, bk.Client())
+	require.NoError(t, err)
+
+	honestManager.Start(honestCtx)
+	evilManager.Start(evilCtx)
+
+	t.Run("crashes mid-challenge and recovers to complete it", func(t *testing.T) {
```

Comment:
OPTIONAL: Consider making these two functions well-named test helper functions just to make the method slightly more concise and to reduce the level of indentation by one.

I think it could be easy when scanning this test method to miss the fact that there are basically two scenarios being tested sequentially after some initial setup.